### DISCUSSION TITLE
Primary assets disappears from asset library but are still linked in compositions and risk analysis

### DISCUSSION THREAD

**Discussion by acherifi** [monarc-project/MonarcAppFO#516] [Q&A]
**Describe the bug**

I was browsing my primary assets to edit their composition then saw a message telling me something in the line of "This asset does not exist" and then I saw a number of primary assets disappear from my asset lists. 
However those assets are still present in the compositions and the risk analysis. 

The problem is that I cannot add them back to the list and cannot search them in the composition search box. 

I know have phantom primary assets in the risk analysis that I cannot modify and perform actions on. 

Is there any way to get them back ? 

**To Reproduce**
Sorry But I cannot remember the precise steps

**Screenshots**
![image](https://github.com/monarc-project/MonarcAppFO/assets/19311305/d4413d1c-71c8-4c86-9b76-d4c976fdbf01)

We can clearly see here that the Sales primary asset category do not have any other asset inside but the asset itself has two other primary assets inside and in the risk analysis

**Desktop (please complete the following information):**

| Questions                                | Answers
|-----------------------------------------|--------------------
| Type of issue                           | Bug
| OS version (server)                 | Ubuntu, ...
| OS version (client)                   | Ubuntu 
| PHP version                             | The one given with the monarc provided VM
| MONARC version / git hash     | 2.12.7
| Browser                                    | Chrome

**Comment by ruslanbaidan** [comment]
The objects composition in the Library can be made of objects from different categories and when the primary asset is added to the analysis the tree is displayed like in the composition hierarchy.
Could that be the case that your assets "Processes to draft offers ..." and "Method and processes ..." are located in another categories  of the library ?

**Reply by acherifi** [reply]
No it is not possible. We designed the assets organization to be the same in the risk analysis than in the library.  I "fixed it" by recreating the assets and redid the composition. 
However the same thing happened a few hours after. Fortunately I took a snapshot that I was able to restore. It is very annoying since it seems like assets are truly deleted from the databse since they are not searchable anymore

**Reply by acherifi** [reply]
Is there any logs I can consult to check the error ?

**Reply by acherifi** [reply]
Interestingly, restoring the snapshot restored the old assets that disappeared. 
I know have the assets twice in the Asset list.

**Comment by ruslanbaidan** [comment]
> Is there any logs I can consult to check the error ?

The case is that there was no visual error, so it will not be presented in the apache error.log

**Comment by ruslanbaidan** [comment]
> Interestingly, restoring the snapshot restored the old assets that disappeared. I know have the assets twice in the Asset list.

In this case I'd assume there was no asset removal, but ether the library display issue nether accident category change.
If there is a bug in the library data result (response from the backend), we need to know the exact scenario to reproduce the same.

**Comment by acherifi** [comment]
I will to try to get as much information as possible if it ever happens again

**Comment by acherifi** [comment]
It happened again, I was modifying primary assets and got this error "this object is not bound to the ANR". And recently modified assets disappeared. I am stuck because I do not have a recent snapshot so I cannot recover it. Also I cannot perform snapshots anymore due to this error "An error occurred while fetching api/client-anr/22/snapshot: Detached entity Monarc\FrontOffice\Model\Entity\Anr@13098 cannot be removed (500)"

Here is the full error 
```
{
    "errors": [
        {
            "message": "Detached entity Monarc\\FrontOffice\\Model\\Entity\\Anr@13098 cannot be removed",
            "error": "error-exception",
            "exception": {
                "class": "Doctrine\\ORM\\ORMInvalidArgumentException",
                "file": "\/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/doctrine\/orm\/lib\/Doctrine\/ORM\/ORMInvalidArgumentException.php",
                "line": 177,
                "message": "Detached entity Monarc\\FrontOffice\\Model\\Entity\\Anr@13098 cannot be removed",
                "stacktrace": "#0 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/doctrine\/orm\/lib\/Doctrine\/ORM\/UnitOfWork.php(2090): Doctrine\\ORM\\ORMInvalidArgumentException::detachedEntityCannot()\n#1 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/doctrine\/orm\/lib\/Doctrine\/ORM\/UnitOfWork.php(2041): Doctrine\\ORM\\UnitOfWork-\u003EdoRemove()\n#2 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/doctrine\/orm\/lib\/Doctrine\/ORM\/EntityManager.php(689): Doctrine\\ORM\\UnitOfWork-\u003Eremove()\n#3 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/monarc\/core\/src\/Model\/Table\/AnrTable.php(74): Doctrine\\ORM\\EntityManager-\u003Eremove()\n#4 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/monarc\/frontoffice\/src\/Service\/AnrService.php(1561): Monarc\\Core\\Model\\Table\\AnrTable-\u003EdeleteEntity()\n#5 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/monarc\/frontoffice\/src\/Service\/SnapshotService.php(63): Monarc\\FrontOffice\\Service\\AnrService-\u003EduplicateAnr()\n#6 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/monarc\/frontoffice\/src\/Controller\/ApiAnrAbstractController.php(124): Monarc\\FrontOffice\\Service\\SnapshotService-\u003Ecreate()\n#7 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-mvc\/src\/Controller\/AbstractRestfulController.php(460): Monarc\\FrontOffice\\Controller\\ApiAnrAbstractController-\u003Ecreate()\n#8 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-mvc\/src\/Controller\/AbstractRestfulController.php(422): Laminas\\Mvc\\Controller\\AbstractRestfulController-\u003EprocessPostData()\n#9 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-eventmanager\/src\/EventManager.php(320): Laminas\\Mvc\\Controller\\AbstractRestfulController-\u003EonDispatch()\n#10 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-eventmanager\/src\/EventManager.php(178): Laminas\\EventManager\\EventManager-\u003EtriggerListeners()\n#11 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-mvc\/src\/Controller\/AbstractController.php(105): Laminas\\EventManager\\EventManager-\u003EtriggerEventUntil()\n#12 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-mvc\/src\/Controller\/AbstractRestfulController.php(306): Laminas\\Mvc\\Controller\\AbstractController-\u003Edispatch()\n#13 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-mvc\/src\/DispatchListener.php(117): Laminas\\Mvc\\Controller\\AbstractRestfulController-\u003Edispatch()\n#14 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-eventmanager\/src\/EventManager.php(320): Laminas\\Mvc\\DispatchListener-\u003EonDispatch()\n#15 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-eventmanager\/src\/EventManager.php(178): Laminas\\EventManager\\EventManager-\u003EtriggerListeners()\n#16 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/vendor\/laminas\/laminas-mvc\/src\/Application.php(319): Laminas\\EventManager\\EventManager-\u003EtriggerEventUntil()\n#17 \/var\/lib\/monarc\/releases\/MonarcAppFO-v2.12.7\/public\/index.php(34): Laminas\\Mvc\\Application-\u003Erun()\n#18 {main}"
            }
        }
    ]
}
``` 

What can I do now ?

**Reply by ruslanbaidan** [reply]
There is a fix to avoid the 1st issue. For this it's necessary to deploy a new patch and a new VM.
To migrate your analysis to the new VM, it's possible to export from the current one and import in the new the whole analysis.

Related to the snapshot creation.
The error log that you sent doesn't provide the exact place of the error, because when something happens in the snapshot creation steps it tries to remove the created data and the error message tells about the failure on removal, but not the root cause.

In general it looks like a missing link between a composed object and analysis that causes the problem.

**Reply by acherifi** [reply]
Indeed It works but this is not a suitable solution. We can't afford to redeploy monarc each time this error occurs. 

It seems that editing an asset impact, saving it and immediately checking the asset in the library by using the "show asset in library button" cause this error. I can reproduce it a 100 times

**Reply by ruslanbaidan** [reply]
The error is occurred when data of the library object is fetched, so [the following fix](https://github.com/monarc-project/zm-core/compare/master...patch-object-display?quick_pull=1) should help assuming that there is a missing link.
The link is not really needed, however some other parts of the code could rely on the link.

If you would like to try if this fix will work for you, you may try to manually modify the file and if the application will work without issues, we can deploy a new patch and VM.

**Reply by acherifi** [reply]
Thank you very much for your reactivity, I highly appreciate it. 

I am now unable to reproduce this error so it might work on the long run. 

Thank you very much

**Reply by ruslanbaidan** [reply]
You are welcome.

Could you tell if the issue is fixed after the code fix is applied ?

Does the Snapshot creation work for you ?

Thank you.

**Reply by acherifi** [reply]
I can now use the "Find asset in library" button. My assets does not disappear anymore.
However the Snapshot function still does not work with the same error as before

**Reply by ruslanbaidan** [reply]
Would it be possible to have have a dump of your monarc_cli database, that we can investigate and fix the issue ?
It could be sent to our email privately and we will remove all the data after finishing the investigation.

The other thing to try with your analysis is to create a new one based on it. There is an option on a new analysis creation `Source -> Existing risk analysis`.
If this action is not failed, as Snapshot creation (could be, as uses the same part of the code), then you could rename sensitive information and export the data of the analysis.

**Reply by acherifi** [reply]
I would love to but unfortunately the problem happened again... My assets disappeared after I performed an import of assets trough a CSV file. I can still "find" my asset by using the "Show asset in the library" button, but the asset tree itself does not show it

**Reply by ruslanbaidan** [reply]
Was it the case before also with an imported asset ?

**Reply by acherifi** [reply]
I don't think it was the case before. However I can't remember clearly so I am not 100% sure.

**Reply by acherifi** [reply]
I tried creating a Risk Analysis from my existing one as you mentionned and it is failing. I get an error when Monarc tries to deleted the attempted creation of the Risk analysis

**Reply by ruslanbaidan** [reply]
This is expected because the same logic is used as for Snapshot creation.
Could you also try to export whole your analysis, create a new one and import into it.
The library compositions are reproduced only for the instantiated library objects (those that are added to the analysis). So if some of them are not there, you need to add before export.

**Reply by acherifi** [reply]
I tried and I cannot do that because of a failed constraints on the unicity of some keys. 
However if I import it in a brand new instance of Monarc it works flawlessly
